### DISCUSSION TITLE
How to read ASDF files from C++ code?

### DISCUSSION THREAD

**Discussion by BenWibking** [spacetelescope/pyasdf#1832] [Q&A] [answered]
It looks like this library is Python-only, at least for now. Is there a recommended way to read ASDF files from C++ code?

**Comment by braingram** [comment, marked as answer]
At the moment the most mature ASDF implementation that I know of for C++ is https://github.com/eschnett/asdf-cxx
Does that look like it will fit your use case? I'm curious to know either way.

**Reply by BenWibking** [reply]
Yes, that looks like it will work.

**Comment by nden** [comment]
@BenWibking Could you briefly describe your use case for a C/C++ library. Are you interfacing with other C/C++ code and if so which one? Ar eyou using data analysis tools in C/C++ and if so which ones? 
This will help with planning development. Thanks!

**Reply by BenWibking** [reply]
The use case is to read data tables that are used by our simulation code (https://github.com/quokka-astro/quokka). We currently use HDF5 for this, but it's overly-complicated and we'd like to include a library that can be built as part of a CMake superbuild of our code, rather than requiring the user to install an external library. I considered just using plain YAML, but the ability to embed binary arrays using ASDF is the main reason I want to use ASDF over YAML.

In the future, it would also be useful to use ASDF to write data tables while our simulation code runs, such as histograms of the simulation state. We currently write those out in YAML or CSV. Again, the main attraction of using ASDF is to save n-dimensional arrays in a portable, self-describing binary format.

We have no plans to do data analysis in C++.
